### ISSUE TITLE
支持SSH escape sequences

### ISSUE THREAD

**Issue by openqt** [trzsz/trzsz-ssh#17]
```
$ ~?
Supported escape sequences:
 ~.   - terminate session
 ~B   - send a BREAK to the remote system
 ~R   - request rekey
 ~#   - list forwarded connections
 ~?   - this message
 ~~   - send the escape character by typing it twice
(Note that escapes are only recognized immediately after newline.)

```
建议支持ssh的一些特殊字符功能，必要的时候还是很好用的。

**Comment by lonnywong** [comment]
我以前没怎么用过，看起来挺有意思的。
我感觉原生的交互做的不太好，我们可以做的更好用一些，欢迎一起来共建。

**Comment by serious-angel** [comment]
> I haven't used it much before, but it looks quite interesting.

Thank you very much for the marvel...

May I ask, how would you terminate the current session without killing the process manually otherwise, if not the shortcut like in the case of `openssh`?

I see there's the following shortcut, but it doesn't seem like anything happens if you press `Ctrl+C Ctrl+Q` to "Quit/Exit" mentioned at <https://trzsz.github.io/ssh>

**Comment by lonnywong** [comment]
> May I ask, how would you terminate the current session without killing the process manually otherwise, if not the shortcut like in the case of `openssh`?

It can be achieved by rewriting the function: https://github.com/trzsz/trzsz-ssh/blob/ca95033bf5d4d157e6b5201d6c7f3eb903b429b6/tssh/trzsz.go#L56
Intercepts the user input (without forwarding it to the server), and blocks the server output (without forwarding it to the user), until the user exits the control interface.

I can take some time to implement an example first, to make it easier for everyone to submit pull requests to add more features.

**Comment by lonnywong** [comment]
Some features are already supported, Check the [document](https://github.com/trzsz/trzsz-ssh/blob/main/README.en.md#ssh-console).

**Comment by serious-angel** [comment]
> Some features are already supported, Check the [document](https://github.com/trzsz/trzsz-ssh/blob/main/README.en.md#ssh-console).

I saw the option, but it didn't work - regardless how much you wait, the sequence `Enter, ~` results in the character `~` sent only.

It's expected to handle the even within the second:

https://github.com/trzsz/trzsz-ssh/blob/b183bea695523a01e283bc18259d321cf1fc8b5a/tssh/trzsz.go#L150

Yet, it still outputs the `~` on the remote, or `~.` for the sequence `Enter, ~, .`.
I've tried explicitly setting the option in `~/.ssh/config`:

```sh
Host example
    Hostname example.com
    Port 1234
    User remote
    IdentityFile /home/shade/.ssh/keys/example
    RequestTTY yes
    # Temporary workaround for chrooted environment on the remote.
    RemoteCommand bash -lc '\. /data/env.sh && bash -c "tmux new -As 0 -- bash"'
    #!! UdpMode yes
    #!! TsshdPath /data/.local/bin/tsshd 
    #!! ConsoleEscapeTime 3
```

Regardless, the same happens, unfortunately.
Thank you for the note, @lonnywong !

**Comment by lonnywong** [comment]
@serious-angel Thank you for your feedback.

Have you installed or compiled the tssh from the latest source code and use it?
Try doing it like this:

```sh
go install github.com/trzsz/trzsz-ssh/cmd/tssh@main
~/go/bin/tssh example
```

The new tssh should be compatible with the old tsshd, but you will see a warning at the top of the screen. I suggest you recompile tsshd as well:
```sh
go install github.com/trzsz/tsshd/cmd/tsshd@main
cp ~/go/bin/tsshd /data/.local/bin/tsshd
```

**Comment by serious-angel** [comment]
> Have you installed or compiled the tssh from the latest source code and use it?

Yes, indeed! The versions are the latest locally and remotely, downloaded from the release pages, and the same issue happens:
```sh
$ tssd --version;
trzsz ssh 0.1.23
$ tsshd --version;
trzsz sshd 0.1.5
```

I tried compiling the latest Go package sources, but the same, too.

**Comment by lonnywong** [comment]
I haven't released a new release yet. The latest release doesn't support it; only the latest code supports it.

Note that after the `@` character is `main`, not `latest`:
```
go install github.com/trzsz/trzsz-ssh/cmd/tssh@main
go install github.com/trzsz/tsshd/cmd/tsshd@main
```

**Comment by serious-angel** [comment]
> ...is `main`, not `latest`

I am very sorry, now noticed...

I Git-cloned and built the source to have the binaries in the source directory instead:
```sh
$ mkdir ./tsshd && cd ./tsshd &&
git clone -b main --depth 1 -- 'https://github.com/trzsz/tsshd' . &&
cd ./cmd/tsshd &&
go build . &&
./tsshd --version;
trzsz sshd 0.1.5
```

It worked! Though the character `.` after `~` did not disconnect, the TUI menu is shown, and termination happens on the 3 option selected!

The terminal is `konsole 25.08.1`.
The remote Tmux is `v3.5a`.

---

<img width="672" height="331" alt="Image" src="https://github.com/user-attachments/assets/5dc18d21-0a1a-491d-b13c-7916453299e8" />

---

This is freaking **awesome**, dear @lonnywong ! 💫✨

**Comment by lonnywong** [comment]
Yes, tssh is implemented completely differently from openssh. `Enter` (newline) and `~` will enter the tssh console.
tssh does not support SSH escape sequences like `~.`, but instead supports a TUI, which is what you see.

**Comment by serious-angel** [comment]
> Yes, tssh is implemented completely differently from openssh. `Enter` (newline) and `~` will enter the tssh console. tssh does not support SSH escape sequences like `~.`, but instead supports a TUI, which is what you see.

Roger that. Have you considered adding a shortcut to this TUI menu for the `.` (to select the option 3) in addition to `q`?  

Does it mean that console for SSH (i.e. `~c`; e.g. to add `-L` or remove `-KL` forwardings within session) is not available, at this moment?

**Comment by lonnywong** [comment]
> Roger that. Have you considered adding a shortcut to this TUI menu for the `.` (to select the option 3) in addition to `q`?

Sure, I will add this feature to make it compatible with SSH escape sequences.

> Does it mean that console for SSH (i.e. ~c; e.g. to add -L or remove -KL forwardings within session) is not available, at this moment?

Yes, it's not supported yet.

The tsshd has undergone significant improvements ( #167 ) , and I plan to release a new version soon once it's stable.

I haven't had time to implement the forwardings feature yet, It might be supported in the version after next.

**Comment by lonnywong** [comment]
@serious-angel Now `~.` is supported, and also supports customizing the escape character by `EscapeChar`:
```
Host a
   EscapeChar !

Host b
   EscapeChar ^B
```

**Comment by serious-angel** [comment]
> ...and also supports customizing the escape character by `EscapeChar`:

Thank you, dear @lonnywong ! ✨ 
You are a genius! I hope to better participate in the project in the future, too!

The project you started and implement is a dear miracle!
I do believe it'll get more attention eventually, and become another magnificent marvel!
